---
title: "Tissue culture RNAseq analysis"
author: "Tianyi"
date: "05/29/2022"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
editor_options:
  chunk_output_type: console
---


```{r setup, include = F}
knitr::opts_chunk$set(echo = T)
```

# Tissue culture RNAseq analysis - Comparison between PI3K activator and no activator 

## Load required library

```{r Load dependencies}
suppressMessages(library(dplyr))
suppressMessages(library(DESeq2))
suppressMessages(library(ggplot2))
suppressMessages(library(pheatmap))
suppressMessages(library(umap))
suppressMessages(library(biomaRt))
suppressMessages(library(org.Hs.eg.db))
suppressMessages(library(DOSE))
suppressMessages(library(pathview))
suppressMessages(library(clusterProfiler))
suppressMessages(library(pheatmap))
suppressMessages(library(VennDiagram))
suppressMessages(library(RColorBrewer))
suppressMessages(library(msigdbr))
suppressMessages(library(edgeR))
suppressMessages(library(stringr))
suppressMessages(library(future))
suppressMessages(library(patchwork))
suppressMessages(library(VennDiagram))
suppressMessages(library(GSEABase))
suppressMessages(library(GSVA))

```

## Raw data input

### Data from Hisat2 alignment

```{r initial data input}
cells <- read.delim(file = "/Users/tianyi.li.2/Desktop/STAT_ovary/data/raw_count.txt", skip = 1)
info <- read.csv(file = "/Users/tianyi.li.2/Desktop/STAT_ovary/data/info.csv", header = T, sep = ";")
cells <- cells[,-c(2:6)]
cells[is.na(cells)] <- 0  
colnames(info) <- c("Sample", "ID", "Culture", "Activator", "Follicle")

```

## Rename rows for count matrix

```{r rename row names for count matrix}
re_name <- function(cells) {
  gene <- cells[,1]
  rownames(cells) <- gene
  cells <- cells[,-1]
  return(cells)
}

cells <- re_name(cells)

re_order <- function(info, cells) {
  genomic_idx <- match(info$Sample, colnames(cells))
  genomic_idx
  cells <- cells[ ,genomic_idx]
  all(info$Sample == colnames(cells))
  return(cells)
}

info <- info[-34,]
info <- info %>% filter(Culture != "LN521")
info <- info %>% 
  mutate(culture = ifelse(str_detect(Culture, "Fresh"), "Fresh", "Cultured")) %>% 
  mutate(cultures = paste0(culture,"_",Activator))

info <- info[-c(2, 18:21, 25, 30:31, 26, 36, 38, 33),]

cells <- re_order(info, cells)

```


## Construct dds object

```{r construct dds object}

dds <- DESeqDataSetFromMatrix(countData=cells, colData = info, design = ~ cultures)

```


## Plot library size / how many reads per sample

```{r sequence depth check, fig.height = 15, fig.width = 15}

theme <- theme(text=element_text(size=9),
               axis.text.x=element_text(color="black", angle = 90),
               axis.text.y=element_text(color="black"),
               axis.ticks=element_blank(),
               strip.background = element_rect(colour=NA, fill=NA),
               plot.background = element_blank(),
               panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(),
               panel.border = element_blank(),
               axis.line.x = element_line(linewidth=0.4),
               axis.line.y = element_line(linewidth=0.4),
               panel.background = element_blank(),
               legend.position = "bottom")

info %>% mutate(read = colSums(counts(dds))) %>%
            ggplot(aes(x=Sample, y=read, fill = cultures)) + geom_bar(stat = "identity") + theme +
            ggtitle("Samples Sequenced Read") + xlab("") + scale_fill_manual(values = c("#8dd3c7","#fb8072","#bebada"))

```

## Filter out genes with reads less than 10

```{r low count filtering}

keep <- rowSums(counts(dds) > 1) >= 7
dds <- dds[keep,]

```

## PCA plot using normalized data

### CPM normalization using edgeR cpm function

```{r PCA plot cpm, fig.height = 4, fig.width = 6}
# check how many variance does each PC explain
# summary(pca)$importance[2,] * 100

info$cultures <- relevel(factor(info$cultures), ref = "Fresh_No")
r <- edgeR::cpm(counts(dds), normalized.lib.sizes = T, log = T)
pca <- prcomp(t(r))
df <- cbind(info, pca$x)
contri <- round(summary(pca)$importance[2,] * 100, 2)
ggplot(df) + geom_point(aes(x=PC1, y = PC2, color = cultures), size = 3) + theme +
      ggtitle(paste0("PCA (cpm)")) + scale_color_manual(values = c("#8dd3c7","#fb8072","#bebada")) +
      xlab(sprintf("PC1 (%s%%)", contri[1])) + ylab(sprintf("PC2 (%s%%)", contri[2])) +
  stat_ellipse(aes(PC1, PC2, color = cultures))

ggplotly(p1)

```

### vst normalization method in DESeq2

```{r PCA plot vst, fig.height = 4, fig.width = 6}

r <- vst(dds, blind = F)
pca <- plotPCA(r, intgroup = c("cultures"), returnData = T)
percentVar <- round(100 * attr(pca, "percentVar"))
ggplot(pca, aes(PC1, PC2, color = cultures)) +
      ggtitle(paste0("PCA (vst)")) + #theme +
      geom_point(size = 3) + xlab(paste0("PC1: ", percentVar[1], "% variance")) +
      ylab(paste0("PC2: ", percentVar[2], "% variance")) +
      scale_color_manual(values = c("#8dd3c7","#fb8072","#bebada"))

```


## UMAP visualization using normalization method

```{r UMAP plot, fig.height = 4, fig.width = 6}

r <- edgeR::cpm(counts(dds), normalized.lib.sizes = T, log = T)
set.seed(1)
normalized_counts <- t(r)
umap_results <- umap::umap(normalized_counts, n_neighbors = 10)
umap_plot_df <- data.frame(umap_results$layout) %>%
        tibble::rownames_to_column("Sample") %>%
        dplyr::inner_join(info, by = "Sample")
ggplot(umap_plot_df, aes(x = X1, y = X2, color = cultures)) + geom_point(size = 3 )+
      ggtitle(paste0("UMAP (vst)")) + #theme + 
  scale_color_manual(values = c("#8dd3c7","#fb8072","#bebada"))


```

## Heatmap: hierarchical clustering to view the similarities between technical replicates

```{r Heatmap for hierarchical clustering, fig.height = 4, fig.width = 6}

r <- vst(dds, blind = F)
r_cor <- cor(assay(r), method = "pearson")
df <- as.data.frame(colData(dds)[,c("cultures", "Activator")])
pheatmap(r_cor, scale = "column", show_rownames = F, annotation = df, fontsize = 15,
         clustering_distance_cols = "euclidean", main = paste0("Heatmap (vst)"))

```

## Remove outliers in heatmap

```{r Outliers removed}
# H1 was removed 

info <- info[-33,]
cells <- re_order(info, cells)

dds <- DESeqDataSetFromMatrix(countData=cells, colData = info, design = ~ cultures)

keep <- rowSums(counts(dds) > 1) >= 7
dds <- dds[keep,]

```

### vst normalization method in DESeq2, PCA after outlier removed

```{r PCA plot vst after outlier removed, fig.height = 4, fig.width = 6}

r <- vst(dds, blind = F)
pca <- plotPCA(r, intgroup = c("cultures"), returnData = T)
percentVar <- round(100 * attr(pca, "percentVar"))
ggplot(pca, aes(PC1, PC2, color = cultures)) +
      ggtitle(paste0("PCA (vst)")) + theme +
      geom_point(size = 3) + xlab(paste0("PC1: ", percentVar[1], "% variance")) +
      ylab(paste0("PC2: ", percentVar[2], "% variance")) +
      scale_color_manual(values = c("#8dd3c7","#fb8072","#bebada")) + stat_ellipse()

```

## Perform DE analysis

```{r perform DE analysis}

#dds$cultures <- relevel(dds$cultures, ref = "Fresh_No")
dds <- DESeq(dds)

```

## Get DE analysis results

```{r DE analysis with full dataset}

cul <- c("Cultured_No", "Cultured_Yes")
comparisonALL <- list()
for(i in cul){
    info$cultures <- as.factor(info$cultures)
          if (i %in% levels(info$cultures)){
            print(i)
            comparisonALL[[i]] <- results(dds, 
                                    contrast = c("cultures", i, "Fresh_No"), 
                                    independentFiltering = T, pAdjustMethod = "fdr") %>% 
                                  as.data.frame() %>% 
                                  mutate(geneid = rownames(.))}
    else { print(paste0(p," No result"))}       
}  

# Significant DEGs
comparison01 <- list()
for (i in 1:length(comparisonALL)) {
  comparison01[[names(comparisonALL)[i]]] <- comparisonALL[[i]] %>% filter(padj<0.01)
}

comparison001 <- list()
for (i in 1:length(comparisonALL)) {
  comparison001[[names(comparisonALL)[i]]] <- comparisonALL[[i]] %>% filter(padj<0.001)
}


save(dds, comparison001, comparisonALL, 
     file = "/Users/tianyi.li.2/Desktop/STAT_ovary/results/DESeq2_H1_removed_fresh_vs_cultured_IVA_included_removed_replicates.Rdata")
```


## get gene names from gene id

```{r get gene names and save output}
ensembl <- useMart(biomart = "ensembl",dataset = "hsapiens_gene_ensembl")
output <- getBM(attributes=c("ensembl_gene_id","external_gene_name","entrezgene_id","description"), mart = ensembl)
#output <- read.csv2(file = "/Users/tianyi.li.2/Desktop/Cell_line_bulk_seq_202011/results/mart.csv", sep = ",")

load(file = "/Users/tianyi.li.2/Desktop/STAT_ovary/results/DESeq2_H1_removed_fresh_vs_cultured_IVA_included.Rdata")

get_id <- function(res.sig) {
  res.sig$geneid <- rownames(res.sig)
  final <- merge(res.sig, output, by.x = "geneid", by.y = "ensembl_gene_id", all.x = T, all.y = F)
  return(final)
}

final_all <- lapply(comparisonALL, function(x) {
  x <- get_id(x)
})

final <- lapply(comparison001, function(x) {
  x <- get_id(x)
})

openxlsx::write.xlsx(final, "/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_fresh_cultured_H1_removed_IVA_included_removed_replicates_sig001_20230529.xlsx")

```

## Random forest/decision tree to get the most important genes from DEGs list

```{r random forest & decision tree}

# get normalized counts from DESeq object for random forest classification
df <- counts(dds, normalized = T) %>% as.data.frame()

# Only use DEGs to find out the most important DEGs that separate three groups
degs <- final[[2]]$geneid # For IVA group
degs <- final[[1]]$geneid # For fragmentation group
degs <- final$geneid # for activator group
#degs <- kgn$geneid # for activator group
df <- df[rownames(df) %in% degs, ]

# Transpose the dataset, sample should be row and column should be DEGs
d1 <- t(df) %>% scale(center = T, scale = T) %>% as.data.frame() %>% 
  mutate(id = rownames(.))
info <- info %>% select(Sample, cultures)
info <- info_activator %>% select(Sample, Activator) # for activator comparison

d1 <- merge(d1, info, by.x = "id", by.y = "Sample", all.x = T, all.y = F)
rownames(d1) <- d1$id
d1 <- d1[, -1]

d2 <- d1 %>% filter(cultures != "Cultured_Yes")
d2 <- d1 # for activator comparison

# split data into testing and training
set.seed(100)
inTest <- createDataPartition(y = d2$cultures, p = 0.25, list = F)
inTest <- createDataPartition(y = d2$Activator, p = 0.25, list = F) # for activator comparison

test <- d2[inTest,]
train <- d2[-inTest,]

# random forest
train$cultures <- factor(train$cultures)
test$cultures <- factor(test$cultures)

train$Activator <- factor(train$Activator)
test$Activator <- factor(test$Activator)

class_forest <- cforest(cultures ~ ., data=train, controls = cforest_control(ntree = 50))
pred <- predict(class_forest, newdata = test, OOB = TRUE)
conf_matrix <- table(pred, test$cultures)
accuracy <- sum(diag(conf_matrix))/sum(conf_matrix)
accuracy

# random forest for activator comparison
class_forest <- cforest(Activator ~ ., data=train, controls = cforest_control(ntree = 5))
class_forest
pred <- predict(class_forest, newdata = test, OOB = TRUE)
conf_matrix <- table(pred, test$Activator)
accuracy <- sum(diag(conf_matrix))/sum(conf_matrix)
accuracy

# Plot the first 4 trees
for (i in 1:5) {
  pt <- prettytree(class_forest@ensemble[[i]], names(class_forest@data@get("input")))
  nt <- new("BinaryTree")
  nt@tree <- pt
  nt@data <- class_forest@data
  nt@responses <- class_forest@responses
  plot(nt, main = paste0("Tree ", i, " from the forest"))
}

# decision tree
train$cultures <- factor(train$cultures)
test$cultures <- factor(test$cultures)

class_tree <- tree(cultures ~ ., data = train)
plot(class_tree)
pred_full <- predict(class_tree, test)
pred_short <- ifelse(pred_full[,"Cultured_No"] > 0.5, "Cultured_No", "Fresh_No")
conf_matrix <- table(pred_short, test$cultures)
conf_matrix

# decision tree with rpart package
class_tree <- rpart(cultures ~ ., train)
rpart.plot(class_tree, box.palette = "RdBu", shadow.col = "gray")

# decision tree for activator comparison
class_tree <- tree(Activator ~ ., data = train)
class_tree
pred_full <- predict(class_tree, test)
pred_short <- ifelse(pred_full[,"Yes"] > 0.5, "Yes", "No")
conf_matrix <- table(pred_short, test$Activator)
conf_matrix

class_tree <- rpart(Activator ~ ., data = train)
rpart.plot(class_tree, box.palette = "RdBu", shadow.col = "gray")

```

## Venn diagram from three group comparison

```{r Venn diagram}

# scale_color_manual(values = c("#8dd3c7","#fb8072","#bebada"))

read_excel_allsheets <- function(filename, tibble = FALSE) {
    sheets <- readxl::excel_sheets(filename)
    x <- lapply(sheets, function(X) readxl::read_excel(filename, sheet = X))
    if(!tibble) x <- lapply(x, as.data.frame)
    names(x) <- sheets
    x
}

kgn <- read_excel_allsheets("/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_fresh_cultured_H1_removed_IVA_included_sig01_20220714.xlsx")
d5 <- kgn[[1]]
d6 <- as.data.frame(intersect(d5$geneid, comparison001[[1]]$geneid))

cols <- c("#8dd3c7","#fb8072")
p1 <- VennDiagram::venn.diagram(
  x = list(na.omit(comparison001[[1]]$geneid), na.omit(comparison001[[2]]$geneid)),
  category.names = c("Fragmentation" , "IVA"),
  #filename = 'Venn for fragmentation and IVA vs fresh overlap sig001.png',
  filename = NULL,
  output=TRUE,
  # Circles
        lwd = 2,
        col=c("#8dd3c7","#fb8072"),
        fill = c(alpha("#8dd3c7",1),alpha("#fb8072",1)),
        scaled = F,
        cex = 1.5,
        fontface = "bold",
        fontfamily = "sans",
        # Set names
        cat.cex = 0.8,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.fontfamily = "sans",
        cat.col =  c("#8dd3c7","#fb8072")
)
ggsave(p1, file="Venn for fragmentation and IVA vs fresh overlap sig001.svg", device = "svg")

```


## Heatmap: DEGs expression

```{r heatmap for selected gene expression, fig.height = 15, fig.width = 6}

stat_df <- final[[2]] %>% dplyr::arrange(desc(.$padj)) %>% dplyr::distinct(.$geneid, .keep_all = T)

normalized_counts <- counts(dds, normalized=T) %>%
  data.frame() %>%
  tibble::rownames_to_column(var="gene") %>%
  as_tibble() %>%
  left_join(output, by=c("gene" = "ensembl_gene_id"))

id_k <- match(stat_df$geneid, normalized_counts$gene)
stat_count <- normalized_counts[id_k,]

genes_k <- stat_count[,(ncol(stat_count)-2)]

scaleSTAT <- scale(t(stat_count[,2:(ncol(stat_count)-4)]), scale = T, center = T) %>% t() %>%
  as.data.frame() %>% cbind(., genes_k) %>%
  cbind(., stat_df$external_gene_name) %>%
  na.omit()

anno <- info %>% select(Sample, cultures, Follicle) %>% filter(cultures != "Cultured_No") 
anno <- info %>% select(Sample, Culture, cultures) 
rownames(anno) <- anno$Sample
STAT_scale <- scaleSTAT[,1:(ncol(scaleSTAT)-2)]

genomic_idx <- match(anno$Sample, colnames(STAT_scale))
genomic_idx
STAT_scale <- STAT_scale[,na.omit(genomic_idx)]
anno <- anno[,-1]

pheatmap(STAT_scale, cluster_rows = T, show_rownames = F,
         cluster_cols = T, clustering_distance_rows = "euclidean" , annotation = anno,
         breaks = seq(-2, 2, length.out = 90))


```

## Heatmap for glycolysis genes

```{r Glycolysis gene heatmap}

glyco <- read_excel_allsheets("/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_fresh_sigDEGs_IVA_included_removed_replicates_GSEA_hallmark_20230529.xlsx")
glyco <- glyco[[2]]

# Get individual genes from enriched gene sets and assign the group and hallmark id
tmp1 <- str_split(glyco[11,]$core_enrichment, "/") %>% as.data.frame() %>% 
        'colnames<-' ("gene")

tmp1 <- str_split(glyco[3,]$core_enrichment, "/") %>% as.data.frame() %>% 
        'colnames<-' ("gene")
tmp1[31:33, 1] <- c("IL2", "IL31", "TNF")

gly <- normalized_counts[which(normalized_counts$external_gene_name %in% tmp1$gene),]

# steroid genes
gly <- normalized_counts[which(normalized_counts$external_gene_name %in% c("CYP11A1", "CYP17A1", "CYP19A1", "STAR", "HSD3B1",
                                                                           "HSD3B2", "AKR1C3", "SRD5A2", "SRD5A1", 
                                                                           "HSD11B1", "HSD11B2", "HSD17B1", "HSD17B3", "HSD17B5",
                                                                           "DHCR7", 
                                                                           "NR5A1" ,"INHA")),]

genes_k <- gly[,(ncol(gly)-2)]

scaleSTAT <- scale(t(gly[,2:(ncol(gly)-3)]), scale = T, center = T) %>% t() %>%
  as.data.frame() %>% cbind(., genes_k) %>%
  cbind(., gly$external_gene_name) %>%
  na.omit()

anno <- info %>% select(Sample, cultures) 
anno <- anno[,2] %>% as.data.frame()
rownames(anno) <- info$Sample
colnames(anno) <- "culture"
STAT_scale <- scaleSTAT[,1:(ncol(scaleSTAT)-2)]
rownames(STAT_scale) <- scaleSTAT$external_gene_name

pheatmap(STAT_scale, cluster_rows = T, show_rownames = T,
         cluster_cols = T, clustering_distance_rows = "euclidean", annotation_col = anno,
         breaks = seq(-1, 1, length.out = 90))

fresh <- STAT_scale[, colnames(STAT_scale) %in% info[info$cultures == "Fresh_No",]$Sample]
fre <- apply(fresh, 1, mean) %>% as.data.frame() %>% 
  'colnames<-' ("Fresh")

frag <- STAT_scale[, colnames(STAT_scale) %in% info[info$cultures == "Cultured_No",]$Sample]
frag <- apply(frag, 1, mean) %>% as.data.frame() %>% 
  'colnames<-' ("Fragmentation")

iva <- STAT_scale[, colnames(STAT_scale) %in% info[info$cultures == "Cultured_Yes",]$Sample]
iva <- apply(iva, 1, mean) %>% as.data.frame() %>% 
  'colnames<-' ("IVA")

d1 <- cbind(fre, frag)
d1 <- cbind(d1, iva)

pheatmap(d1, cluster_rows = T, show_rownames = T,
         cluster_cols = F, clustering_distance_rows = "euclidean",
         breaks = seq(-0.8, 0.8, length.out = 90))

```

## Volcano plot

```{r volcano, fig.height = 6, fig.width = 15}
pval_threshold <- 0.01
logfc_threshold <- 1

deseq.results <- comparisonALL
deseq.results <- deseq.results %>% filter(!is.na(padj))
deseq.results$threshold <- as.factor(abs(deseq.results$log2FoldChange) >= logfc_threshold & 
                             deseq.results$padj < pval_threshold)
deseq.results <- merge(deseq.results, output[,2:3], by.x = "geneid", by.y = "ensembl_gene_id", 
                             all.x = T, all.y = F)
ggplot(data=deseq.results, 
            aes(x=log2FoldChange, y=-log10(padj), colour=threshold)) +
            geom_point(alpha=0.4, size=1.75) +
            theme(legend.position = "none") +
            theme_bw() + theme(legend.position="none") +
            geom_vline(xintercept = logfc_threshold) +
            geom_vline(xintercept = -logfc_threshold) +
            geom_hline(yintercept = -log10(pval_threshold)) +
            xlab("log2 fold change") + ylab("-log10 FDR") +
            ggtitle("DEGs") +
            ggrepel::geom_text_repel(aes(label=ifelse(padj < pval_threshold & abs(log2FoldChange) >= logfc_threshold,
                                   external_gene_name, '')))


```


## Shrunk results for downstream functional analysis

```{r shrunk results}
# coef input should be like: "group_DES_10.6M_vs_DMSO"

dds$cultures <- relevel(dds$cultures, ref = "Fresh_No")
dds <- nbinomWaldTest(dds, maxit=1000)

shrunk <- list()
for(i in cul) {
  p <- paste0("cultures_", i, "_vs_Fresh_No")
    if (p %in% resultsNames(dds)) {
        print(p)
        shrunk[[i]] <- lfcShrink(dds, coef = p, type = "apeglm") %>%
        as.data.frame() %>% mutate(geneid = rownames(.))
    } else { print(paste0(p," No result"))} 
}

shrunk_sig <- list()
for (i in 1:length(shrunk)) {
    shrunk_sig[[names(shrunk)[i]]] <- shrunk[[i]] %>% filter(padj<0.001)
}

shrunk_sig <- list()
for (i in 1:length(shrunk)) {
    shrunk_sig[[names(shrunk)[i]]] <- shrunk[[i]] %>% filter(padj<0.01)
}

final_shrunk <- lapply(shrunk_sig, function(x) {
  x <- get_id(x)
})

all_shrunk <- lapply(shrunk, function(x) {
  x <- get_id(x)
})

save(shrunk, shrunk_sig, 
     file = "/Users/tianyi.li.2/Desktop/STAT_ovary/results/DESeq2_H1_removed_IVA_included_fresh_cultured_shrunk.Rdata")
```

## Fragmentation and IVA specific DEGs enrichment

```{r specific DEGs enrichment}

common <- intersect(final_shrunk[[1]]$geneid, final_shrunk[[2]]$geneid)
fra <- final_shrunk[[1]]$geneid[!final_shrunk[[1]]$geneid %in% common]
iv <- final_shrunk[[2]]$geneid[!final_shrunk[[2]]$geneid %in% common]

fragment <- final_shrunk[[1]][final_shrunk[[1]]$geneid %in% fra,]
IVA <- final_shrunk[[2]][final_shrunk[[2]]$geneid %in% iv,]

genelist <- IVA %>% 
    dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    mutate(logpadj = ifelse(log2FoldChange < 0, log10(padj), -log10(padj))) %>% 
    filter(!is.na(entrezgene_id))
GeneList <- genelist$log2FoldChange
names(GeneList) <- as.character(genelist$entrezgene_id)
GeneList <- sort(GeneList, decreasing = T)

gsea <- GSEA(gene = GeneList, 
                   TERM2GENE = msigHs_h,
                   pvalueCutoff = 1,
                   pAdjustMethod = "BH",
                   minGSSize = 10,
                   maxGSSize = 500)

gsea <- setReadable(gsea, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
openxlsx::write.xlsx(gsea, "/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/GSEA_IVA_specific_hallmark_IVA_included_remove_replicates_20230529.xlsx")

enrich <- enricher(gene = IVA$entrezgene_id, 
                   TERM2GENE = msigHs_h,
                   pvalueCutoff = 0.1,
                   pAdjustMethod = "BH",
                   minGSSize = 10,
                   maxGSSize = 500)
enrich <- setReadable(enrich, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
openxlsx::write.xlsx(enrich, "/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/enrich_IVA_specific_hallmark_IVA_included_20230112.xlsx")

```


## Get background gene list

-log10(padj) then get the sign of the log2FC. If upregulated, positive. If downregulated, negative. Then rank according to the logpadj. The nonsignificant ones will end up in the middle (which is not important in the enrichement)
Almost similar to ranking with log2FC

```{r get background genes for functional analysis}

genelist <- all_shrunk %>% 
    dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    mutate(logpadj = ifelse(log2FoldChange < 0, log10(padj), -log10(padj))) %>% 
    filter(!is.na(entrezgene_id))
GeneList <- genelist$log2FoldChange
names(GeneList) <- as.character(genelist$entrezgene_id)
GeneList <- sort(GeneList, decreasing = T)


# Significant genes only

GeneList <- list()
for(i in 1:length(final_shrunk)){
  genelist <- final_shrunk[[i]] %>% 
    dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    mutate(logpadj = ifelse(log2FoldChange < 0, log10(padj), -log10(padj))) %>% 
    filter(!is.na(entrezgene_id))
  GeneList[[names(final_shrunk)[i]]] <- genelist$log2FoldChange
  names(GeneList[[names(final_shrunk)[i]]]) <- as.character(genelist$entrezgene_id)
  GeneList[[names(final_shrunk)[i]]] <- sort(GeneList[[names(final_shrunk)[i]]], decreasing = T)
}

```

## Msigdb gene set analysis

```{r Msigdb gene set analysis}

msigHs_h <- msigdbr(species = "Homo sapiens", category = "H") %>%
  dplyr::select(gs_name, entrez_gene) %>% 
  as.data.frame()


enrich <- enricher(gene = all_shrunk$entrezgene_id, 
                   TERM2GENE = msigHs_h,
                   pvalueCutoff = 0.1,
                   pAdjustMethod = "BH",
                   minGSSize = 10,
                   maxGSSize = 500)
enrich <- setReadable(enrich, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
gse_df <- enrich %>% as.data.frame()
gse_df

gse <- list()
gse_df <- list()
for (i in 1:length(final_shrunk)) {
  print(paste0("Gene set analysis for"," ", names(final_shrunk)[i]," ", "vs Fresh"))
  enrich <- enricher(gene = final_shrunk[[i]]$entrezgene_id, 
                   TERM2GENE = msigHs_h,
                   pvalueCutoff = 0.1,
                   pAdjustMethod = "BH",
                   minGSSize = 10,
                   maxGSSize = 500)
  gse[[names(final_shrunk)[i]]] <- enrich
  gse_df[[names(final_shrunk)[i]]] <- enrich %>% as.data.frame()
}

for (i in 1:length(gse)) {
  gse[[i]] <- setReadable(gse[[i]], OrgDb = org.Hs.eg.db, keyType="ENTREZID")
}

openxlsx::write.xlsx(gse, "/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_fresh_sigDEGs_enricher_hallmark_IVA_included_20220714.xlsx")

gsea_hallmark <- list()
gsea_hallmark_df <- list()
for (i in 1:length(GeneList)){
  print(names(GeneList)[i])
  gsea <- GSEA(gene = GeneList[[i]], 
                   TERM2GENE = msigHs_h,
                   pvalueCutoff = 1,
                   pAdjustMethod = "BH",
                   minGSSize = 10,
                   maxGSSize = 500)
  gsea_hallmark[[names(GeneList)[i]]] <- gsea
  gsea_hallmark_df[[names(GeneList)[i]]] <- gsea %>% as.data.frame() 
}

for (i in 1:length(gsea_hallmark)) {
  gsea_hallmark[[i]] <- setReadable(gsea_hallmark[[i]], OrgDb = org.Hs.eg.db, keyType="ENTREZID")
}

openxlsx::write.xlsx(gsea_hallmark, "/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_fresh_sigDEGs0.01_IVA_included_removed_replicates_GSEA_hallmark_20230529.xlsx")

```

## Barplot for GSEA analysis 

```{r barplot for GSEA results}

gse <- read_excel_allsheets("/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_fresh_sigDEGs_IVA_included_removed_replicates_GSEA_hallmark_20230529.xlsx")

gse <- gse[1:2]

frag <- gse[[1]] %>% filter(p.adjust < 0.1) %>% 
  select(ID, NES) %>% 
  mutate(group = rep("Fragmentation", nrow(.)))
iva <- gse[[2]] %>% filter(p.adjust < 0.1) %>% 
  select(ID, NES) %>% 
  mutate(group = rep("IVA", nrow(.)))

iva <- iva %>% arrange(desc(NES))
frag <- frag[match(iva$ID[-c(12, 15, 18, 20, 21)], frag$ID),]

gse_all <- rbind(iva, frag)
gse_all <- gse_all %>% mutate(regulate = ifelse(str_detect(NES, "-"), "Down", "Up"))
gse_all$score <- abs(gse_all$NES)
gse_all$score[23:nrow(gse_all)] <- paste0("-", gse_all$score[23:nrow(gse_all)]) %>% as.numeric()
gse_all$ID <- factor(gse_all$ID, levels = iva$ID)

ggplot(gse_all, aes(x = ID, y = score, fill = regulate)) + 
  geom_bar(position = "dodge", stat = "identity") +
  geom_hline(yintercept = 0, color = "black", size = 1) +
  coord_flip() + ylab("NES") + xlab("") +
  ggtitle("GSEA analysis") +
  theme(text=element_text(size=15),
               axis.text.x=element_text(color="black", angle = 90, vjust = 0.3, hjust = 0.3),
               axis.text.y=element_text(color="black"),
               axis.ticks.x.bottom=element_blank(),
               #axis.ticks.length.x=unit(-.25, "cm"),
               strip.background = element_rect(colour=NA, fill=NA),
               plot.background = element_blank(),
               #panel.grid.major = element_blank(),
               #panel.grid.minor = element_blank(),
               panel.spacing.x=unit(.5, "lines"),
               panel.border = element_rect(color = "black", fill = NA, size = 1),
               axis.line.x = element_line(size=0.4),
               axis.line.y = element_line(size=0.4),
               legend.key.size = unit(2, 'lines'),
               panel.background = element_blank()
)

act <- read_excel_allsheets("/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_Activator_all_DEGs_GSEA_hallmark_20230529.xlsx")
act <- act[[1]] %>% arrange(desc(NES))
act <- act %>% mutate(regulate = ifelse(str_detect(NES, "-"), "Down", "Up"))
act$score <- abs(act$NES)
act$ID <- factor(act$ID, levels = act$ID)

ggplot(act, aes(x = ID, y = score, fill = regulate)) + 
  geom_bar(position = "dodge", stat = "identity") +
  coord_flip() + ylab("NES") + xlab("") +
  ggtitle("GSEA analysis") +
  theme(text=element_text(size=15),
               axis.text.x=element_text(color="black", angle = 90, vjust = 0.3, hjust = 0.3),
               axis.text.y=element_text(color="black"),
               axis.ticks.x.bottom=element_blank(),
               #axis.ticks.length.x=unit(-.25, "cm"),
               strip.background = element_rect(colour=NA, fill=NA),
               plot.background = element_blank(),
               #panel.grid.major = element_blank(),
               #panel.grid.minor = element_blank(),
               panel.spacing.x=unit(.5, "lines"),
               panel.border = element_rect(color = "black", fill = NA, size = 1),
               axis.line.x = element_line(size=0.4),
               axis.line.y = element_line(size=0.4),
               legend.key.size = unit(2, 'lines'),
               panel.background = element_blank()
)

```

## Dotplot for Hallmark using enricher and gsea

```{r Dotplot, fig.height=20, fig.width=10}

cultured_no <- gsea_hallmark[[1]]
dotplot(gsea_hallmark[[1]], showCategory=22, orderBy="GeneRatio") + labs(title="gsea_hallmark")


dotplot(gse[[2]], showCategory=26, orderBy="GeneRatio") + labs(title="enricher_hallmark cultured yes")

```

## Emapp plot for H

```{r emap plot for Hallmark, fig.height=10, fig.width=10}

test <- enrichplot::pairwise_termsim(gsea)
emapplot(test, showCategory = 22) + ggtitle("Emap plot - GSEA")

test <- enrichplot::pairwise_termsim(enrich)
emapplot(test, showCategory = 22) + ggtitle("Emap plot - enricher")

```

## enrichKEGG and gseKEGG

```{r enrich KEGG}

enrich_KEGG <- enrichKEGG(gene = all_shrunk$entrezgene_id, 
                       organism = 'hsa',
                       pvalueCutoff = 0.1)

enrich_KEGG <- enrichKEGG(gene = final_shrunk$entrezgene_id, 
                       organism = 'hsa',
                       pvalueCutoff = 0.1)

enrich_KEGG <- setReadable(enrich_KEGG, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
enrich_KEGG_df <- enrich_KEGG %>% as.data.frame() 

openxlsx::write.xlsx(enrich_KEGG_df, "/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_H1_removed_sigDEGs_enricher_KEGG_20220706.xlsx")

########## Plot selected KEGG across groups ###########

#selected <- c("hsa04350", "hsa04928", "hsa04150", "hsa04310", "hsa04022", "hsa04024",
#              "hsa04914", "hsa04114", "hsa04914", "hsa04919", "hsa00062", "hsa00071", 
#              "hsa01212", "hsa01040", "hsa01212", "hsa04024", "hsa04910", "hsa04020",
#              "hsa04010", "hsa04151")

edo <- gseKEGG(geneList = GeneList, 
                organism     = 'hsa',
               nPerm        = 1000,
               minGSSize    = 10,
               pvalueCutoff = 0.1,
               verbose      = FALSE)
edo <- setReadable(edo, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
gse_KEGG_df <- edo %>% as.data.frame()

openxlsx::write.xlsx(gse_KEGG_df, "/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_A6_E3_removed_gse_KEGG_20220529.xlsx")

```

## GO analysis

```{r GO analysis}

ego <- enrichGO(gene = final_shrunk$entrezgene_id, 
                                                         universe = all_shrunk$entrezgene_id,
                                                         OrgDb = org.Hs.eg.db, ont = "ALL", 
                                                         pAdjustMethod = "BH", 
                                                         pvalueCutoff = 0.05, readable = TRUE)
ego <- setReadable(ego, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
enrich_GO_df <- ego %>% as.data.frame()

openxlsx::write.xlsx(enrich_GO_df, "/Users/tianyi.li.2/Desktop/STAT_ovary/results/tables/DESeq2_H1_removed_fresh_enrich_GO_20220708.xlsx")

```

## Plot GO analysis results

```{r plot GO analysis results}

dotplot(ego, showCategory = 50,  orderBy="GeneRatio") +
      ggtitle("Dotplot - enrich GO") + scale_color_continuous(low = "purple", high = "green") 


```


```{r Session info}
sessionInfo()
```
