---
title: "Tissue culture RNAseq analysis"
author: "Tianyi"
date: "05/29/2022"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_float: yes
  pdf_document:
    toc: yes
editor_options:
  chunk_output_type: console
---


```{r setup, include = F}
knitr::opts_chunk$set(echo = T)
```

# Tissue culture RNAseq analysis - Comparison between PI3K activator and no activator 

## Load required library

```{r Load dependencies}
suppressMessages(library(dplyr))
#suppressMessages(library(tximport))
#suppressMessages(library(S4Vectors))
#suppressMessages(library(readr))
#suppressMessages(library(EnsDb.Hsapiens.v86))
#suppressMessages(library(ensembldb))
suppressMessages(library(DESeq2))
suppressMessages(library(ggplot2))
suppressMessages(library(pheatmap))
suppressMessages(library(umap))
suppressMessages(library(biomaRt))
suppressMessages(library(org.Hs.eg.db))
suppressMessages(library(DOSE))
suppressMessages(library(pathview))
suppressMessages(library(clusterProfiler))
suppressMessages(library(pheatmap))
suppressMessages(library(VennDiagram))
suppressMessages(library(RColorBrewer))
suppressMessages(library(msigdbr))
suppressMessages(library(edgeR))
suppressMessages(library(stringr))
suppressMessages(library(future))
suppressMessages(library(patchwork))
suppressMessages(library(VennDiagram))
suppressMessages(library(GSEABase))
suppressMessages(library(GSVA))

```

## Raw data input

### Data from Hisat2 alignment

```{r initial data input}
cells <- read.delim(file = "/Users/tili/Desktop/STAT_ovary/data/raw_count.txt", skip = 1)
info <- read.csv(file = "/Users/tili/Desktop/STAT_ovary/data/info.csv", header = T, sep = ";")
cells <- cells[,-c(2:6)]
cells[is.na(cells)] <- 0  
colnames(info) <- c("Sample", "ID", "Culture", "Activator", "Follicle")

```

## Rename rows for count matrix

```{r rename row names for count matrix}
re_name <- function(cells) {
  gene <- cells[,1]
  rownames(cells) <- gene
  cells <- cells[,-1]
  return(cells)
}

cells <- re_name(cells)

re_order <- function(info, cells) {
  genomic_idx <- match(info$Sample, colnames(cells))
  genomic_idx
  cells <- cells[ ,genomic_idx]
  all(info$Sample == colnames(cells))
  return(cells)
}

info <- info[-34,]
info <- info %>% filter(Activator != "Yes") %>%
  mutate(culture = ifelse(str_detect(Culture, "Fresh"), "Fresh", "Cultured"))
info <- info %>% filter(Culture != "LN521")

cells <- re_order(info, cells)

```


## Construct dds object

```{r construct dds object}

dds <- DESeqDataSetFromMatrix(countData=cells, colData = info, design = ~ culture)

```


## Plot library size / how many reads per sample

```{r sequence depth check, fig.height = 15, fig.width = 15}

theme <- theme(text=element_text(size=9),
               axis.text.x=element_text(color="black", angle = 90),
               axis.text.y=element_text(color="black"),
               axis.ticks=element_blank(),
               strip.background = element_rect(colour=NA, fill=NA),
               plot.background = element_blank(),
               panel.grid.major = element_blank(),
               panel.grid.minor = element_blank(),
               panel.border = element_blank(),
               axis.line.x = element_line(size=0.4),
               axis.line.y = element_line(size=0.4),
               panel.background = element_blank(),
               legend.position = "bottom")

info %>% mutate(read = colSums(counts(dds))) %>%
            ggplot(aes(x=Sample, y=read, fill = culture)) + geom_bar(stat = "identity") + theme +
            ggtitle("Samples Sequenced Read") + xlab("") + scale_fill_manual(values = c("#8dd3c7","#fb8072"))

```

## Filter out genes with reads less than 10

```{r low count filtering}

keep <- rowSums(counts(dds) > 1) >= 7
dds<- dds[keep,]

```

## PCA plot using normalized data

### CPM normalization using edgeR cpm function

```{r PCA plot cpm, fig.height = 4, fig.width = 6}
# check how many variance does each PC explain
# summary(pca)$importance[2,] * 100

info$culture <- relevel(factor(info$culture), ref = "Fresh")
r <- edgeR::cpm(counts(dds), normalized.lib.sizes = T, log = T)
pca <- prcomp(t(r))
df <- cbind(info, pca$x)
contri <- round(summary(pca)$importance[2,] * 100, 2)
ggplot(df) + geom_point(aes(x=PC1, y = PC2, color = culture), size = 3) + theme +
      ggtitle(paste0("PCA (cpm)")) + scale_color_manual(values = c("#8dd3c7","#fb8072")) +
      xlab(sprintf("PC1 (%s%%)", contri[1])) + ylab(sprintf("PC2 (%s%%)", contri[2])) +
  stat_ellipse(aes(PC1, PC2, color = culture))

ggplotly(p1)

```

### vst normalization method in DESeq2

```{r PCA plot vst, fig.height = 4, fig.width = 6}

r <- vst(dds, blind = F)
pca <- plotPCA(r, intgroup = c("culture"), returnData = T)
percentVar <- round(100 * attr(pca, "percentVar"))
ggplot(pca, aes(PC1, PC2, color = culture)) +
      ggtitle(paste0("PCA (vst)")) + theme +
      geom_point(size = 3) + xlab(paste0("PC1: ", percentVar[1], "% variance")) +
      ylab(paste0("PC2: ", percentVar[2], "% variance")) +
      scale_color_manual(values = c("#8dd3c7","#fb8072"))

```


## UMAP visualization using normalization method

```{r UMAP plot, fig.height = 4, fig.width = 6}

r <- edgeR::cpm(counts(dds), normalized.lib.sizes = T, log = T)
set.seed(1)
normalized_counts <- t(r)
umap_results <- umap::umap(normalized_counts, n_neighbors = 10)
umap_plot_df <- data.frame(umap_results$layout) %>%
        tibble::rownames_to_column("Sample") %>%
        dplyr::inner_join(info, by = "Sample")
ggplot(umap_plot_df, aes(x = X1, y = X2, color = culture)) + geom_point(size = 3 )+
      ggtitle(paste0("UMAP (vst)")) + theme + scale_color_manual(values = c("#fb8072", "#8dd3c7"))


```

## Heatmap: hierarchical clustering to view the similarities between technical replicates

```{r Heatmap for hierarchical clustering, fig.height = 4, fig.width = 6}

r <- vst(dds, blind = F)
r_cor <- cor(assay(r), method = "pearson")
df <- as.data.frame(colData(dds)[,c("culture", "Activator")])
pheatmap(r_cor, scale = "column", show_rownames = F, annotation = df, fontsize = 15,
         clustering_distance_cols = "euclidean", main = paste0("Heatmap (vst)"))

```

## Remove outliers in heatmap

```{r Outliers removed}
# H1 was removed 

info <- info[-23,]
cells <- re_order(info, cells)

dds <- DESeqDataSetFromMatrix(countData=cells, colData = info, design = ~ culture)

keep <- rowSums(counts(dds) > 1) >= 7
dds <- dds[keep,]

```

### vst normalization method in DESeq2, PCA after outlier removed

```{r PCA plot vst after outlier removed, fig.height = 4, fig.width = 6}

r <- vst(dds, blind = F)
pca <- plotPCA(r, intgroup = c("culture"), returnData = T)
percentVar <- round(100 * attr(pca, "percentVar"))
ggplot(pca, aes(PC1, PC2, color = culture)) +
      ggtitle(paste0("PCA (vst)")) + theme +
      geom_point(size = 3) + xlab(paste0("PC1: ", percentVar[1], "% variance")) +
      ylab(paste0("PC2: ", percentVar[2], "% variance")) +
      scale_color_manual(values = c("#8dd3c7","#fb8072")) + stat_ellipse()

```

## Perform DE analysis

```{r perform DE analysis}

dds <- DESeq(dds)

```

## Get DE analysis results

```{r DE analysis with full dataset}

comparisonALL <- results(dds, contrast = c("culture", "Cultured", "Fresh"), 
                                  independentFiltering = T, pAdjustMethod = "fdr") %>% 
                                  as.data.frame() %>% 
                                  mutate(geneid = rownames(.))

# Significant DEGs
comparison05 <- comparisonALL %>% filter(padj < 0.05)
comparison01 <- comparisonALL %>% filter(padj < 0.01)

save(dds, comparison01, comparisonALL, 
     file = "/Users/tili/Desktop/STAT_ovary/results/DESeq2_H1_removed_fresh_vs_cultured.Rdata")
```


## get gene names from gene id

```{r get gene names and save output}
#ensembl <- useMart(biomart = "ensembl",dataset = "hsapiens_gene_ensembl")
#output <- getBM(attributes=c("ensembl_gene_id","external_gene_name","entrezgene_id","description"), mart = ensembl)
output <- read.csv2(file = "/Users/tili/Desktop/Cell_line_bulk_seq_202011/results/mart.csv", sep = ",")

load(file = "/Users/tili/Desktop/STAT_ovary/results/DESeq2_H1_removed_fresh_vs_cultured.Rdata")

get_id <- function(res.sig) {
  res.sig$geneid <- rownames(res.sig)
  final <- merge(res.sig, output, by.x = "geneid", by.y = "ensembl_gene_id", all.x = T, all.y = F)
  return(final)
}

final_all <- get_id(comparisonALL)
final <- get_id(comparison01)

openxlsx::write.xlsx(final, "/Users/tili/Desktop/STAT_ovary/results/tables/DESeq2_fresh_cultured_H1_removed_sig01_20220708.xlsx")

```


## Heatmap: DEGs expression

```{r heatmap for selected gene expression, fig.height = 15, fig.width = 6}

stat_df <- final %>% arrange(desc(.$padj)) %>% dplyr::distinct(.$geneid, .keep_all = T)

normalized_counts <- counts(dds, normalized=T) %>%
  data.frame() %>%
  tibble::rownames_to_column(var="gene") %>%
  as_tibble() %>%
  left_join(output, by=c("gene" = "ensembl_gene_id"))

id_k <- match(stat_df$geneid, normalized_counts$gene)
stat_count <- normalized_counts[id_k,]

genes_k <- stat_count[,(ncol(stat_count)-2)]

scaleSTAT <- scale(t(stat_count[,2:(ncol(stat_count)-4)]), scale = T, center = T) %>% t() %>%
  as.data.frame() %>% cbind(., genes_k) %>%
  cbind(., stat_df$external_gene_name) %>%
  na.omit()

anno <- info %>% select(culture, Follicle)
rownames(anno) <- info$Sample
STAT_scale <- scaleSTAT[,1:(ncol(scaleSTAT)-2)]

pheatmap(STAT_scale, cluster_rows = T, show_rownames = F,
         cluster_cols = T, clustering_distance_rows = "correlation", annotation = anno,
         breaks = seq(-2, 2, length.out = 90))


```


## Volcano plot

```{r volcano, fig.height = 6, fig.width = 15}
pval_threshold <- 0.01
logfc_threshold <- 1

deseq.results <- comparisonALL
deseq.results <- deseq.results %>% filter(!is.na(padj))
deseq.results$threshold <- as.factor(abs(deseq.results$log2FoldChange) >= logfc_threshold & 
                             deseq.results$padj < pval_threshold)
deseq.results <- merge(deseq.results, output[,2:3], by.x = "geneid", by.y = "ensembl_gene_id", 
                             all.x = T, all.y = F)
ggplot(data=deseq.results, 
            aes(x=log2FoldChange, y=-log10(padj), colour=threshold)) +
            geom_point(alpha=0.4, size=1.75) +
            theme(legend.position = "none") +
            theme_bw() + theme(legend.position="none") +
            geom_vline(xintercept = logfc_threshold) +
            geom_vline(xintercept = -logfc_threshold) +
            geom_hline(yintercept = -log10(pval_threshold)) +
            xlab("log2 fold change") + ylab("-log10 FDR") +
            ggtitle("DEGs") +
            ggrepel::geom_text_repel(aes(label=ifelse(padj < pval_threshold & abs(log2FoldChange) >= logfc_threshold,
                                   external_gene_name, '')))


```


## Shrunk results for downstream functional analysis

```{r shrunk results}
# coef input should be like: "group_DES_10.6M_vs_DMSO"

dds$culture <- relevel(dds$culture, ref = "Fresh")
dds <- nbinomWaldTest(dds, maxit=1000)
shrunk <- lfcShrink(dds, coef =  "culture_Cultured_vs_Fresh", type = "apeglm") %>%
        as.data.frame() %>% mutate(geneid = rownames(.))

shrunk_sig <- shrunk %>% filter(padj<0.01)

final_shrunk <- get_id(shrunk_sig)

all_shrunk <- get_id(shrunk)

save(shrunk, shrunk_sig, 
     file = "/Users/tili/Desktop/STAT_ovary/results/DESeq2_H1_removed_fresh_cultured_shrunk.Rdata")
```

## Get background gene list

-log10(padj) then get the sign of the log2FC. If upregulated, positive. If downregulated, negative. Then rank according to the logpadj. The nonsignificant ones will end up in the middle (which is not important in the enrichement)
Almost similar to ranking with log2FC

```{r get background genes for functional analysis}

genelist <- all_shrunk %>% 
    dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    mutate(logpadj = ifelse(log2FoldChange < 0, log10(padj), -log10(padj))) %>% 
    filter(!is.na(entrezgene_id))
GeneList <- genelist$log2FoldChange
names(GeneList) <- as.character(genelist$entrezgene_id)
GeneList <- sort(GeneList, decreasing = T)


# Significant genes only
genelist <- final_shrunk %>% 
    dplyr::select(entrezgene_id, log2FoldChange, padj) %>% 
    mutate(logpadj = ifelse(log2FoldChange < 0, log10(padj), -log10(padj))) %>% 
    filter(!is.na(entrezgene_id))
GeneList <- genelist$log2FoldChange
names(GeneList) <- as.character(genelist$entrezgene_id)
GeneList <- sort(GeneList, decreasing = T)


```

## Msigdb gene set analysis

```{r Msigdb gene set analysis}

msigHs_c2 <- msigdbr(species = "Homo sapiens", category = "C2") %>%
  dplyr::select(gs_name, entrez_gene) %>% 
  as.data.frame()

msigHs_h <- msigdbr(species = "Homo sapiens", category = "H") %>%
  dplyr::select(gs_name, entrez_gene) %>% 
  as.data.frame()


enrich <- enricher(gene = all_shrunk$entrezgene_id, 
                   TERM2GENE = msigHs_h,
                   pvalueCutoff = 0.1,
                   pAdjustMethod = "BH",
                   minGSSize = 10,
                   maxGSSize = 500)
enrich <- setReadable(enrich, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
gse_df <- enrich %>% as.data.frame()
gse_df

enrich <- enricher(gene = final_shrunk$entrezgene_id, 
                   TERM2GENE = msigHs_h,
                   pvalueCutoff = 0.1,
                   pAdjustMethod = "BH",
                   minGSSize = 10,
                   maxGSSize = 500)
enrich <- setReadable(enrich, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
gse_df <- enrich %>% as.data.frame()
gse_df

openxlsx::write.xlsx(gse_df, "/Users/tili/Desktop/STAT_ovary/results/tables/DESeq2_fresh_sigDEGs_enricher_hallmark_20220708.xlsx")

gsea <- GSEA(gene = GeneList, 
                   TERM2GENE = msigHs_h,
                   pvalueCutoff = 0.1,
                   pAdjustMethod = "BH",
                   minGSSize = 10,
                   maxGSSize = 500)
gsea <- setReadable(gsea, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
gsea_hallmark_df <- gsea %>% as.data.frame() 
gsea_hallmark_df

openxlsx::write.xlsx(gsea, "/Users/tili/Desktop/STAT_ovary/results/tables/DESeq2_fresh_sigDEGs_GSEA_hallmark_20220708.xlsx")

```

## Dotplot for Hallmark using enricher and gsea

```{r Dotplot, fig.height=20, fig.width=10}

dotplot(gsea, showCategory=22, orderBy="GeneRatio") + labs(title="gsea_hallmark")

dotplot(enrich, showCategory=26, orderBy="GeneRatio") + labs(title="enricher_hallmark")

```

## Emapp plot for H

```{r emap plot for Hallmark, fig.height=10, fig.width=10}

test <- enrichplot::pairwise_termsim(gsea)
emapplot(test, showCategory = 22) + ggtitle("Emap plot - GSEA")

test <- enrichplot::pairwise_termsim(enrich)
emapplot(test, showCategory = 22) + ggtitle("Emap plot - enricher")

```

## enrichKEGG and gseKEGG

```{r enrich KEGG}

enrich_KEGG <- enrichKEGG(gene = all_shrunk$entrezgene_id, 
                       organism = 'hsa',
                       pvalueCutoff = 0.1)

enrich_KEGG <- enrichKEGG(gene = final_shrunk$entrezgene_id, 
                       organism = 'hsa',
                       pvalueCutoff = 0.1)

enrich_KEGG <- setReadable(enrich_KEGG, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
enrich_KEGG_df <- enrich_KEGG %>% as.data.frame() 

openxlsx::write.xlsx(enrich_KEGG_df, "/Users/tili/Desktop/STAT_ovary/results/tables/DESeq2_H1_removed_sigDEGs_enricher_KEGG_20220706.xlsx")

########## Plot selected KEGG across groups ###########

#selected <- c("hsa04350", "hsa04928", "hsa04150", "hsa04310", "hsa04022", "hsa04024",
#              "hsa04914", "hsa04114", "hsa04914", "hsa04919", "hsa00062", "hsa00071", 
#              "hsa01212", "hsa01040", "hsa01212", "hsa04024", "hsa04910", "hsa04020",
#              "hsa04010", "hsa04151")

edo <- gseKEGG(geneList = GeneList, 
                organism     = 'hsa',
               nPerm        = 1000,
               minGSSize    = 10,
               pvalueCutoff = 0.1,
               verbose      = FALSE)
edo <- setReadable(edo, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
gse_KEGG_df <- edo %>% as.data.frame()

openxlsx::write.xlsx(gse_KEGG_df, "/Users/tili/Desktop/STAT_ovary/results/tables/DESeq2_A6_E3_removed_gse_KEGG_20220529.xlsx")

```

## GO analysis

```{r GO analysis}

ego <- enrichGO(gene = final_shrunk$entrezgene_id, 
                                                         universe = all_shrunk$entrezgene_id,
                                                         OrgDb = org.Hs.eg.db, ont = "ALL", 
                                                         pAdjustMethod = "BH", 
                                                         pvalueCutoff = 0.05, readable = TRUE)
ego <- setReadable(ego, OrgDb = org.Hs.eg.db, keyType="ENTREZID")
enrich_GO_df <- ego %>% as.data.frame()

openxlsx::write.xlsx(enrich_GO_df, "/Users/tili/Desktop/STAT_ovary/results/tables/DESeq2_H1_removed_fresh_enrich_GO_20220708.xlsx")

```

## Plot GO analysis results

```{r plot GO analysis results}

dotplot(ego, showCategory = 50,  orderBy="GeneRatio") +
      ggtitle("Dotplot - enrich GO") + scale_color_continuous(low = "purple", high = "green") 


```


```{r Session info}
sessionInfo()
```
